substrate-node/pallets/pallet-tfgrid/src/farm.rs

```diff
@@ -1,10 +1,373 @@
+    ensure,
+    sp_runtime::SaturatedConversion,
+    traits::Get,
+    BoundedVec, RuntimeDebug,
+};
 use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
 use scale_info::TypeInfo;
-use sp_std::{marker::PhantomData, vec::Vec};
+use sp_std::{marker::PhantomData, vec, vec::Vec};
+use tfchain_support::{
+    traits::ChangeNode,
+    types::{Farm, FarmCertification, FarmingPolicyLimit, PublicIP, IP4},
+};
 
-use crate::{Config, Error, FarmNameInput};
+impl<T: Config> Pallet<T> {
+    pub fn _create_farm(
+        account_id: T::AccountId,
+        name: FarmNameInput<T>,
+        public_ips: PublicIpListInput<T>,
+    ) -> DispatchResultWithPostInfo {
+        let mut id = FarmID::<T>::get();
+        id = id + 1;
+
+        let twin_id = TwinIdByAccountID::<T>::get(&account_id).ok_or(Error::<T>::TwinNotExists)?;
+        let twin = Twins::<T>::get(twin_id).ok_or(Error::<T>::TwinNotExists)?;
+        ensure!(
+            twin.account_id == account_id,
+            Error::<T>::CannotCreateFarmWrongTwin
+        );
+
+        ensure!(
+            !FarmIdByName::<T>::contains_key(name.clone()),
+            Error::<T>::FarmExists
+        );
+        let farm_name = Self::get_farm_name(name.clone())?;
+
+        let public_ips_list = Self::get_public_ips(public_ips)?;
+
+        let new_farm = Farm {
+            version: TFGRID_FARM_VERSION,
+            id,
+            twin_id,
+            name: farm_name,
+            pricing_policy_id: 1,
+            certification: FarmCertification::NotCertified,
+            public_ips: public_ips_list,
+            dedicated_farm: false,
+            farming_policy_limits: None,
+        };
+
+        Farms::<T>::insert(id, &new_farm);
+        FarmIdByName::<T>::insert(name.to_vec(), id);
+        FarmID::<T>::put(id);
+
+        Self::deposit_event(Event::FarmStored(new_farm));
+
+        Ok(().into())
+    }
+
+    pub fn _update_farm(
+        account_id: T::AccountId,
+        farm_id: u32,
+        name: FarmNameInput<T>,
+    ) -> DispatchResultWithPostInfo {
+        let new_farm_name = Self::get_farm_name(name.clone())?;
+
+        let twin_id = TwinIdByAccountID::<T>::get(&account_id).ok_or(Error::<T>::TwinNotExists)?;
+
+        let mut farm = Farms::<T>::get(farm_id).ok_or(Error::<T>::FarmNotExists)?;
+
+        ensure!(
+            farm.twin_id == twin_id,
+            Error::<T>::CannotUpdateFarmWrongTwin
+        );
+
+        if FarmIdByName::<T>::contains_key(name.clone()) {
+            let farm_id_by_new_name = FarmIdByName::<T>::get(name.clone());
+            // if the user picks a new name but it is taken by another farmer, don't allow it
+            if farm_id_by_new_name != farm_id {
+                return Err(Error::<T>::InvalidFarmName.into());
+            }
+        }
+
+        let name: Vec<u8> = farm.name.into();
+        // Remove stored farm by name and insert new one
+        FarmIdByName::<T>::remove(name.clone());
```

Comment:
Hmm, following the logic it s done on purpose
But I see an issue here because the idea is to remove the entry with `old` farm name and create a new entry with `new` farm name
What is done here is re-creating an entry with `old` farm name

Reply:
Oh. But I assume changing the name doesn't happen too often... Let's fix this indeed